File: client/src/api/endpoints/recurring.ts
import type { AxiosInstance } from 'axios'
import type { RecurringFrequency, RecurringRule, RecurringStatus, UpcomingBill } from '../../types'

export interface RecurringRuleCreateInput {
  name: string
  amount: number
  frequency: RecurringFrequency
  next_due: string
  category_id?: string | null
}

export interface RecurringRuleUpdateInput {
  name?: string
  amount?: number
  frequency?: RecurringFrequency
  next_due?: string
  category_id?: string | null
  status?: RecurringStatus
}

export interface PostRecurringRuleInput {
  budget_id: string
  date?: string
}

export const recurringApi = {
  list: (client: AxiosInstance) => client.get<RecurringRule[]>('/api/recurring-rules').then((res) => res.data),
  upcoming: (client: AxiosInstance, days?: number) =>
    client.get<UpcomingBill[]>('/api/recurring-rules/upcoming', { params: days ? { days } : undefined }).then((res) => res.data),
  create: (client: AxiosInstance, payload: RecurringRuleCreateInput) =>
    client.post<RecurringRule>('/api/recurring-rules', payload).then((res) => res.data),
  update: (client: AxiosInstance, id: string, payload: RecurringRuleUpdateInput) =>
    client.patch<RecurringRule>(`/api/recurring-rules/${id}`, payload).then((res) => res.data),
  remove: (client: AxiosInstance, id: string) => client.delete(`/api/recurring-rules/${id}`).then(() => undefined),
  post: (client: AxiosInstance, id: string, payload: PostRecurringRuleInput) =>
    client.post<RecurringRule>(`/api/recurring-rules/${id}/post`, payload).then((res) => res.data),
}

File: server/app/schemas/recurring.py
import datetime as dt
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel

RecurringFrequency = Literal["weekly", "monthly", "quarterly", "yearly"]
RecurringStatus = Literal["active", "paused"]


class RecurringRuleCreate(BaseModel):
    name: str
    amount: Decimal
    frequency: RecurringFrequency
    next_due: dt.date
    category_id: str | None = None


class RecurringRuleUpdate(BaseModel):
    name: str | None = None
    amount: Decimal | None = None
    frequency: RecurringFrequency | None = None
    next_due: dt.date | None = None
    category_id: str | None = None
    status: RecurringStatus | None = None


class RecurringRuleResponse(BaseModel):
    """Shape expected by the frontend's `RecurringRule` type (client/src/types/index.ts)."""

    id: str
    name: str
    amount: Decimal
    frequency: RecurringFrequency
    nextDue: str
    category: str
    status: RecurringStatus


class PostRecurringRuleRequest(BaseModel):
    """Posts a due recurring rule into a real transaction on the given budget."""

    budget_id: str
    date: dt.date | None = None


class UpcomingBillResponse(BaseModel):
    """Shape expected by the frontend's `UpcomingBill` type — derived from recurring rules."""

    id: str
    name: str
    dueDate: str
    amount: Decimal
    category: str
